shinysdr/plugins/flightradar24/__init__.py

```diff
@@ -0,0 +1,267 @@
+_FEET_PER_MINUTE_TO_METERS_PER_SECOND = _METERS_PER_FEET * 60
+
+
+# TODO: This really shouldn't be a Device, but that's the only way right now to hook into the config.
+def Flightradar24(reactor, key='flightradar24', bounds=None):
+    """Create a flightradar24 client.
+
+    key: Component ID.
+    bounds: optional 4-element tuple of (lat1, lat2, lon1, lon2) to restrict search
+    """
+    return Device(components={six.text_type(key): _Flightradar24Client(
+        reactor=reactor,
+        bounds=bounds)})
+
+
+@implementer(IComponent, ITopWatcher)
+class _Flightradar24Client(ExportedState):
+    def __init__(self, reactor, bounds):
+        self.__reactor = reactor
+        self.__agent = Agent(reactor)
+        self.__bounds = bounds
+        self.__top = None
+        self.__loop = None
+
+    @exported_value(type=bool, changes='this_setter', label='Enabled')
+    def get_enabled(self):
+        return self.__loop is not None
+
+    @setter
+    def set_enabled(self, enabled):
+        if enabled and not self.__loop:
+            self.__loop = task.LoopingCall(self.__send_request)
+            self.__loop.clock = self.__reactor
+            self.__loop.start(_POLLING_INTERVAL).addErrback(print)
+        elif not enabled and self.__loop:
+            self.__loop.stop()
+            self.__loop = None
+
+    def close(self):
+        if self.__loop:
+            self.__loop.stop()
+            self.__loop = None
+
+    def set_top(self, top):
+        self.__top = top
+
+    def __make_url(self):
+        u = URL.fromText('https://data-live.flightradar24.com/zones/fcgi/feed.js?faa=1&mlat=1&flarm=1&adsb=1&gnd=0&air=1&vehicles=0&estimated=1&maxage=14400&gliders=1&stats=0')
+        if self.__bounds:
+            u = u.set('bounds', ','.join(str(b) for b in self.__bounds))
+        return six.binary_type(u.asText())
+
+    def __send_request(self):
+        if not self.__top:
+            return
+        d = self.__agent.request(six.binary_type('GET'), self.__make_url())
+        d.addCallback(readBody)
+        
+        def process(body):
+            data = json.loads(body)
+            for object_id, aircraft in six.iteritems(data):
+                if not isinstance(aircraft, list):
+                    continue
+                self.__top.get_telemetry_store().receive(AircraftWrapper(object_id, aircraft))
+        d.addCallback(process)
+        d.addErrback(print)
+
+
+@implementer(ITelemetryMessage)
+class AircraftWrapper(object):
+    def __init__(self, object_id, message):
+        self.object_id = object_id
+        self.message = message  # list
+    
+    def get_object_id(self):
+        # TODO: add prefix to ensure uniqueness?
+        return self.object_id
+    
+    def get_object_constructor(self):
+        return Aircraft
+
+
+class IAircraft(Interface):
+    """marker interface for client"""
+    pass
+
+
+@implementer(IAircraft, ITelemetryObject)
+class Aircraft(ExportedState):
```

Comment:
This should share an `Aircraft` telemetry object type with the Mode S plugin so that information from both sources can be combined and not duplicated (and we don't need to duplicate the map layer code too).

There's already a multi-plugin situation in APRS — there, there is a plugin for core APRS functionality and a plugin for the multimon-ng decoder. Here, I'd suggest for now depending on the one in the Mode S plugin. If you need to make any changes to it, add comments there that it's used in two places.

The logic to decode the two kinds of messages can go in the message objects.

Reply:
Separate layers can be handled by adding something like a "from what data sources" field to the objects and having the layer provide subordinate filtering options (much like the database layer lets you enter search text). That gets you the functionality without duplicating code and preventing data merging.

(Now that I've thought of it, it seems reasonable to have 'data sources for this object' be a _general_ feature of the telemetry system; e.g. someone might want to filter APRS the same way.)